doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,266 @@
+requires the user to set OSSL_ALG_PARAM_STRICT_CHECKS in their init.
+
+e.g.
+
+```c
+int strict = 0;
+params[0] = OSSL_PARAM_construct_int(OSSL_ALG_PARAM_STRICT_CHECKS, strict);
+EVP_DigestSignInit_ex(ctx, &pctx, name, libctx, NULL, pkey, params);
+// using EVP_PKEY_CTX_set_params() here would be too late
+```
+
+Delaying the check to after the init would be possible, but it would be a change
+in existing behaviour. For example the keysize checks are done in the init since
+this is when the key is setup.
+
+Notes
+----------
+
+There was discussion related to also having a global config setting that could
+turn off FIPS mode. This will not be added at this stage.
+
+New Changes Required
+--------------------
+
+The following changes are required for FIPS 140-3 and will require indicators.
+On a cases by case basis we must decide what to do when unapproved mode is
+detected.
+The mechanism using FIPS configuration options and the indicator callback should
+be used for most of these unapproved cases (rather than always returning an error).
+
+### key size >= 112 bits
+
+There are a few places where we do not enforce key size that need to be addressed.
+
+- HMAC  Which applies to all algorithms that use HMAC also (e.g. HKDF, SSKDF, KBKDF)
+- CMAC
+- KMAC
+
+### Algorithm Transitions
+
+- DES_EDE3_ECB.  Disallowed for encryption, allowed for legacy decryption
+- DSA.  Keygen and Signing are no longer approved, verify is still allowed.
+- ECDSA B & K curves are deprecated, but still approved according to (IG C.K Resolution 4).\
+  If we chose not to remove them , then we need to check that OSSL_PKEY_PARAM_USE_COFACTOR_ECDH is set for key agreement if the cofactor is not 1.
+- ED25519/ED448 is now approved.
+- X25519/X448 is not approved currently. keygen and keyexchange would also need an indicator if we allow it?
+- RSA encryption(for key agreement/key transport) using PKCSV15 is no longer allowed. (Note that this breaks TLS 1.2 using RSA for KeyAgreement),
+  Padding mode updates required. Check RSA KEM also.
+- RSA signing using PKCS1 is still allowed (i.e. signature uses shaXXXWithRSAEncryption)
+- RSA signing using X931 is no longer allowed. (Still allowed for verification). Check if PSS saltlen needs a indicator (Note FIPS 186-4 Section 5.5 bullet(e). Padding mode updates required in rsa_check_padding(). Check if sha1 is allowed?
+- RSA - (From SP800-131Ar2) RSA >= 2048 is approved for keygen, signatures and key transport. Verification allows 1024 also. Note also that according to the (IG section C.F) that fips 186-2 verification is also allowed (So this may need either testing OR an indicator - it also mentions the modulus size must be 1024 * 256*s). Check that rsa_keygen_pairwise_test() and RSA self tests are all compliant with the above RSA restrictions.
+
+- TLS1_PRF  If we are only trying to support TLS1.2 here then we should remove the tls1.0/1.1 code from the FIPS MODULE.
```

Comment:
It's not clear to me what work this is suggesting.

Reply:
Are you talking about line 213?

Reply:
If you are - I am referring to tls1_prf_alg() which has a TLS1.1 path if sha1ctx !=NULL

Reply:
NIST permits TLS 1.1.  Only 1.0 and before are banned.  1.1 is a should not rather than a must not from memory.

Reply:
RSA KEX is AFAIK banned by current NIST rules and creating SHA1 based signatures is banned as well. I do not think you can use TLS-1.1 without at least one of these banned algorithms.

Reply:
TLS1-PRF (https://www.openssl.org/docs/manmaster/man7/EVP_KDF-TLS1_PRF.html) is used for both TLS 1.0, 1.1, and 1.2. For TLS 1.0 and 1.1 specifically, the digest must be set to "MD5-SHA1". However, the fipsprov actually doesn't define MD5, so `EVP_KDF_CTX_set_params` will always fail (assuming TLS1-PRF was fetched with "fips=yes").

Reply:
The FIPS provider doesn't export MD5 but it contains it to support the TLS PRF.

NIST has an exception for this use case (somewhere).  You are allowed to use TLS 1.1 and remain FIPS compliant.

Reply:
@paulidale Yes, NIST allows it (for TLS 1.0/1.1), I'm just describing the OpenSSL 3 FIPS provider behavior. See e.g. https://mta.openssl.org/pipermail/openssl-users/2022-March/015051.html. I don't want to derail this thread, so if you think this behavior is not intended I'll open a separate issue with a proof-of-concept.

Reply:
But we do not allow TLS-1.1 with our fips provider already, so IMO this is moot.